### PR TITLE
Fix Widget #625

### LINKED ISSUE #625: [Bug] Search function with widget do not work properly in specific case.
### Are you ready to report?
    Yes 👍  

### Environment
  * version of the software you use : version name 2.3 (2018-03-12 master branch)
  * device / operating system : Samsung Galaxy S8, Huawei Honor A6
  
### The Bug:

  bug_report

In specific case when I use widget search, It does not go to search screen, goes to latest wiki page.

#### Steps to reproduce:

  1. Click widget's "search kiwix" button
  2. Then search screen will be appear. Press back press button to go to latest wiki page
  3. In latest wiki page, Press home button, go to home.
  4. At home, click widget's "search kiwix" button again.
  5. We'll see latest wiki page, not search screen.

If It work properly, at step 5 we can see search screen. But this case, It does not appear.

**I would like to work this issue.**

### PR DESCRIPTION
**Pull request by prajurock**
Fixes #625 

Screenshots/GIF for the change: [If possible, please add relevant screenshots/GIF]

### CODE REVIEW

**Comment by prajurock** [comment]
@abdulwd please review.

**Comment by prajurock** [comment]
@soloturn @mhutti1 Have a look!!

File: app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java [hunk 1]
```diff
@@ -1304,6 +1308,10 @@ private void updateBottomToolbarVisibility() {
   @Override
   protected void onNewIntent(Intent intent) {
     super.onNewIntent(intent);
+    boolean handled = handleIntent(intent);
```
Comment (abdulwd) [inline app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java]: No need of this variable, this line can be merged with the next.

Comment (prajurock) [reply app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java]: Ok @abdulwd

**Comment by prajurock** [comment]
@abdulwd Done

**Comment by prajurock** [comment]
> would you please add a commit message which makes reviewing easier, as well as reading the commit history afterwards? i tried to google for example why you prefer setAction(null) to setAction("") and did not find.

As far as I know, An empty string is a string instance of zero length, whereas a null string has no value at all. So when we opening widget second time after opening activity in the background it doesn't come to main search screen.

**Comment by soloturn** [comment]
> > would you please add a commit message which makes reviewing easier, as well as reading the commit history afterwards? i tried to google for example why you prefer setAction(null) to setAction("") and did not find.
> 
> As far as I know, An empty string is a string instance of zero length, whereas a null string has no value at all. So when we opening widget second time after opening activity in the background it doesn't come to main search screen.

if such a text would be in the commit message it would be ideal.

**Comment by prajurock** [comment]
> > > would you please add a commit message which makes reviewing easier, as well as reading the commit history afterwards? i tried to google for example why you prefer setAction(null) to setAction("") and did not find.
> > 
> > 
> > As far as I know, An empty string is a string instance of zero length, whereas a null string has no value at all. So when we opening widget second time after opening activity in the background it doesn't come to main search screen.
> 
> if such a text would be in the commit message it would be ideal.

@soloturn okay i will change the commit message.Is there any more changes left @mhutti1 @abdulwd

**Comment by soloturn** [comment]
> > > > would you please add a commit message which makes reviewing easier, as well as reading the commit history afterwards? i tried to google for example why you prefer setAction(null) to setAction("") and did not find.
> > > 
> > > 
> > > As far as I know, An empty string is a string instance of zero length, whereas a null string has no value at all. So when we opening widget second time after opening activity in the background it doesn't come to main search screen.
> > 
> > 
> > if such a text would be in the commit message it would be ideal.
> 
> @soloturn okay i will change the commit message.Is there any more changes left @mhutti1 @abdulwd

you are waiting for one of us @prajurock for the commit message?

**Comment by prajurock** [comment]
@soloturn I was busy. I am changing the commit message.

File: app/src/main/java/org/kiwix/kiwixmobile/main/KiwixSearchWidget.java [hunk 2]
```diff
@@ -49,20 +49,23 @@ public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] a
...
       mainIntent.setAction(TEXT_CLICKED);
+      mainIntent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
       PendingIntent searchPendingIntent =
           PendingIntent.getActivity(context, (int) (System.currentTimeMillis() % Integer.MAX_VALUE),
               mainIntent, 0);
 
       /** Kiwix icon intent to main app **/
       Intent kiwixIconIntent = new Intent(context, MainActivity.class);
       kiwixIconIntent.setAction(ICON_CLICKED);
+      mainIntent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
```
Comment (macgills) [inline app/src/main/java/org/kiwix/kiwixmobile/main/KiwixSearchWidget.java]: Shouldn't this be kwixiIconIntent? also same for starIntent below

File: app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java [hunk 3]
```diff
@@ -1254,20 +1256,20 @@ public void onResume() {
...
           goToSearch(false);
           break;
         case KiwixSearchWidget.STAR_CLICKED:
-          intent.setAction("");
+          intent.setAction(null);
           goToBookmarks();
           break;
         case KiwixSearchWidget.MIC_CLICKED:
-          intent.setAction("");
+          intent.setAction(null);
```
Comment (macgills) [inline app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java]: If the action is just being set to null is there even a point of setting it at all?

**Comment by kelson42** [comment]
@prajurock We would like to carry on with that PR. Is that possible for you to answer last reviewer questions?

File: app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java (hunk 3 above)
Comment (prajurock) [reply app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java]: As far as I know, An empty string is a string instance of zero length, whereas a null string has no value at all. So when we opening widget second time after opening activity in the background it doesn't come to main search screen.

File: app/src/main/java/org/kiwix/kiwixmobile/main/KiwixSearchWidget.java (hunk 2 above)
Comment (prajurock) [reply app/src/main/java/org/kiwix/kiwixmobile/main/KiwixSearchWidget.java]: Voice intent is not working!!

Comment (prajurock) [reply app/src/main/java/org/kiwix/kiwixmobile/main/KiwixSearchWidget.java]: I think error with the widget is that when we open it again then it shows the problem
please check issue you will get to know #625

**Comment by prajurock** [comment]
> 
> 
> @prajurock We would like to carry on with that PR. Is that possible for you to answer last reviewer questions?

Done

File: app/src/main/java/org/kiwix/kiwixmobile/main/KiwixSearchWidget.java (hunk 2 above)
Comment (macgills) [reply app/src/main/java/org/kiwix/kiwixmobile/main/KiwixSearchWidget.java]: This code is still the same code duplicated. I would like to see the code look like this
```
views.setOnClickPendingIntent(R.id.search_widget_text, pendingIntent(context, TEXT_CLICKED));

views.setOnClickPendingIntent(R.id.search_widget_icon, pendingIntent(context, ICON_CLICKED));

views.setOnClickPendingIntent(R.id.search_widget_star, pendingIntent(context, MIC_CLICKED));

views.setOnClickPendingIntent(R.id.search_widget_mic, pendingIntent(context, STAR_CLICKED));
```

File: app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java (hunk 3 above)
Comment (macgills) [reply app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java]: After looking at the code I saw that the source of the intent was `Activity.getIntent()`, weird that we are clearing the intent's action and we are in fact not even saving this as there is no corresponding call to `Activity.setIntent`.
Please delete all the setAction calls. They are most likely an error

Comment (macgills) [reply app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java]: Ah I think I know what it is now, this code is "duplicated" in onResume and in onNewIntent.
It needs to be in onCreate and onNewIntent, both should have a method that handles the intent and depending on the action launches the correct activity

Comment (macgills) [reply app/src/main/java/org/kiwix/kiwixmobile/main/MainActivity.java]: To make the matter more confusing onCreate already has intent handling but not based on the action, instead by extras like onNewIntent.
Even more confusingly I cannot see where these extras get set.

**Comment by kelson42** [comment]
@macgills Are you satisfied?

**Comment by macgills** [comment]
@kelson42 I am not satisfied, there are 2 unresolved conversations, I left a comment on both but forgot to unresolve 1, have done so now.
Also it would be much preferable for this to target develop now.

**Comment by kelson42** [comment]
OK, @prajurock I would really like to see another one from you with the last comments of @macgills taking in account. Hope to see it soon :)

**Pull request closed without merging**